cmd/contour/serve.go

```diff
@@ -206,30 +200,25 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 	// using the SyncList to keep track of what to sync later.
 	var informerSyncList k8s.InformerSyncList
 
-	informerSyncList.Add(dynamicInformerFactory.ForResource(ingressroutev1.IngressRouteGVR).Informer()).AddEventHandler(dynamicHandler)
-	informerSyncList.Add(dynamicInformerFactory.ForResource(ingressroutev1.TLSCertificateDelegationGVR).Informer()).AddEventHandler(dynamicHandler)
-	informerSyncList.Add(dynamicInformerFactory.ForResource(projectcontour.HTTPProxyGVR).Informer()).AddEventHandler(dynamicHandler)
-	informerSyncList.Add(dynamicInformerFactory.ForResource(projectcontour.TLSCertificateDelegationGVR).Informer()).AddEventHandler(dynamicHandler)
+	gvri := k8s.DefaultGVRInformer(dynamicInformerFactory, ctx.UseExperimentalServiceAPITypes)
 
-	informerSyncList.Add(informerFactory.Core().V1().Services().Informer()).AddEventHandler(dynamicHandler)
-	informerSyncList.Add(informerFactory.Networking().V1beta1().Ingresses().Informer()).AddEventHandler(dynamicHandler)
+	// TODO(youngnick): Add in filtering the gvri map by enabled apiserver types (#2219) using the discovery library.
 
-	if ctx.UseExperimentalServiceAPITypes {
-		log.Info("Enabling Experimental Service APIs types")
-		informerSyncList.Add(dynamicInformerFactory.ForResource(serviceapis.GroupVersion.WithResource("gatewayclasses")).Informer()).AddEventHandler(dynamicHandler)
-		informerSyncList.Add(dynamicInformerFactory.ForResource(serviceapis.GroupVersion.WithResource("gateways")).Informer()).AddEventHandler(dynamicHandler)
-		informerSyncList.Add(dynamicInformerFactory.ForResource(serviceapis.GroupVersion.WithResource("httproutes")).Informer()).AddEventHandler(dynamicHandler)
-		informerSyncList.Add(dynamicInformerFactory.ForResource(serviceapis.GroupVersion.WithResource("tcproutes")).Informer()).AddEventHandler(dynamicHandler)
+	for _, inf := range gvri.Informers {
+		informerSyncList.RegisterInformer(inf, dynamicHandler)
 	}
 
+	informerSyncList.RegisterInformer(informerFactory.Core().V1().Services().Informer(), dynamicHandler)
+	informerSyncList.RegisterInformer(informerFactory.Networking().V1beta1().Ingresses().Informer(), dynamicHandler)
```

Comment:
We can get rid of this code if we add the corresponding GVRs to the `InformerSet`

Reply:
We can't without further refactoring, because the core and networking stuff comes from the `informerFactory` (which is not dynamic), and `DefaultInformerSet` currently takes a `DynamicSharedInformerFactory` only. 

I'll look at refactoring this as part of the next round of changes.

Reply:
OK, sounds good. AFAIK, there should not be any problem with using a dynamic informer to watch the core types.

Reply:
I think we'll need some extra unstructured translation, but I'll look at that next.